### PR TITLE
Integrate xds server and sidecar container option into existing test infrastructure.

### PR DESCRIPTION
**Pull request by wanlin31**

### CODE REVIEW

File: containers/init/ready/ready.go [hunk 1]
```diff
@@ -252,6 +257,50 @@ func WaitForReadyPods(ctx context.Context, ltg LoadTestGetter, pl PodLister, tes
 	return podAddresses, &nodesInfo, nil
 }
 
+func passTarget(quitServer bool, clientIP string, targets []*pb.Endpoint, testType string) error {
```
Comment (gnossen) [inline containers/init/ready/ready.go:260]: I don't see any place where `quitServer` is supplied with a value of `false`. Any reason to keep it as a parameter?

File: containers/init/ready/ready.go [hunk 2]
```diff
@@ -252,6 +257,50 @@ func WaitForReadyPods(ctx context.Context, ltg LoadTestGetter, pl PodLister, tes
...
+		log.Fatalf("could not greet: %v", err)
+		statusCode, _ := grpcstatus.FromError(err)
+		log.Print(statusCode.Details()...)
+	}
+	// Used separate goroutines to stop server, non-blocking
+	if quitServer {
+		go func() {
+			fmt.Println("stopping endpoint update server")
+			if _, err := c.QuitTestUpdateServer(ctx, &pb.Void{}); err != nil {
+				statusCode, _ := grpcstatus.FromError(err)
```
Comment (gnossen) [inline containers/init/ready/ready.go:284]: How critical is it if this RPC fails?

File: containers/runtime/xds/README.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  configuration files, the flag here is to make sure that at lease one of the
+  listener resource has the required name.
+
+- -sidecar-listener-port
+
+  This field in only for validation. The listener port that sidecar proxy is
+  listening on. The traffic wish to go through envoy should be send to this
+  port.
+
+- -validation-only
```
Comment (gnossen) [inline containers/runtime/xds/README.md:89]: Suggestion: Maybe call this `-validate-only` instead? Please also add a sentence clarifying that the server will not actually run if this flag is specified. That wasn't clear to me until I read the code.

File: containers/runtime/xds/README.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  The path of default configuration JSON file that the resource snapshot based
+  on. The default value of the field is `config/default_config.json`.
+
+- -custom-config-path
+
+  The path of the user supplied configuration JSON file. No default value for
+  this field, for more information check section:
+  [Custom configuration of xDS server](#Custom-configuration-of-xDS-server)
+
+- -non-proxied-target-string
```
Comment (gnossen) [inline containers/runtime/xds/README.md:74]: I like the extra validation. 👍

File: containers/runtime/xds/cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		l.Errorf("fail to validate the generated snapshot for xDS server: %v", err)
+	}
+
+	l.Infof("xDS server resource snapshot is generated successfully")
+
+	if validationOnly {
+		return
+	}
+	// Move the bootstrap file
+	if pathToBootstrap != "" {
```
Comment (gnossen) [inline containers/runtime/xds/cmd/main.go:83]: Maybe add a comment about _why_ this file is being moved to where it is.

File: containers/runtime/xds/config/config.go [hunk 6]
```diff
@@ -459,3 +453,35 @@ func (t *TestResource) UpdateEndpoint(snap cache.Snapshot) error {
 	}
 	return nil
 }
+
+// SocketListenerOnly takes a pointer of a snapshot, and returns only the socket listeners.
+// This function is used for Proxied test since api_listneners which are used for the
+// non-proxed test can not be validated by Envoy causing the entire resources slices not to be
+// registed.
+func SocketListenerOnly(snap *cache.Snapshot) error {
```
Comment (gnossen) [inline containers/runtime/xds/config/config.go:461]: Nit: Function names should be verb phrases.

File: containers/runtime/xds/logger.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+// Debugf print out debug information.
+func (logger Logger) Debugf(format string, args ...interface{}) {
+	log.Printf(format+"\n", args...)
+}
+
+// Infof print out useful information.
+func (logger Logger) Infof(format string, args ...interface{}) {
+	log.Printf(format+"\n", args...)
```
Comment (gnossen) [inline containers/runtime/xds/logger.go:18]: Will this print a timestamp?

File: kubehelpers/configMap.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+package kubehelpers
+
+import (
+	"encoding/json"
+)
+
+// UpdateConfigMapWithServerPort accepts a server port string and a scenarioString string.
+// It returns a updated scenarioString with server port inserted. Currently only supports
+// single scenario.
+// TODO: wanlin31 to update the function to support array of scenarios.
```
Comment (gnossen) [inline kubehelpers/configMap.go:26]: Missed TODO?

File: kubehelpers/psm.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"fmt"
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+)
+
+// IsPSMTest checks if a given Loadtest is a PSM test. All clients running
+// a scenario should have the same configuration otherwise a error will be
+// returned.
+func IsPSMTest(clients *[]grpcv1.Client) (bool, error) {
```
Comment (gnossen) [inline kubehelpers/psm.go:28]: It kind of feels like this function is doing two things: validating the test spec and then seeing what type of test spec it is.

Maybe split this out into two functions? Then you would only need one place in the code asserting that all clients must run the same kind of test.

File: proto/testupdater/test_updater.proto [hunk 10]
```diff
@@ -16,16 +16,28 @@ syntax = "proto3";
...
-package endpointupdater;
+package testupdater;
 
-service EndpointUpdater {
+service TestUpdater {
   // Sends an update
-  rpc UpdateEndpoint (EndpointUpdaterRequest) returns (EndpointUpdaterReply) {}
+  rpc UpdateTest (TestUpdateRequest) returns (TestUpdateReply) {}
+
+  rpc QuitTestUpdateServer(Void) returns (Void);
```
Comment (gnossen) [inline proto/testupdater/test_updater.proto:25]: Optional: RPCs like this at Google are traditionally called `QuitQuitQuit`

File: status/status.go [hunk 11]
```diff
@@ -54,7 +54,7 @@ func StateForContainerStatus(status *corev1.ContainerStatus) (State, *int32) {
 	if terminateState := status.State.Terminated; terminateState != nil {
 		var state State = Errored
 
-		if terminateState.ExitCode == 0 {
+		if terminateState.ExitCode == 0 || terminateState.ExitCode == 2 {
```
Comment (gnossen) [inline status/status.go:57]: This seems odd. Why is this necessary?

**Review by gnossen — COMMENTED** [review]
Tremendous progress!

File: config/constants.go [hunk 12]
```diff
@@ -66,10 +66,40 @@ const (
 	// DriverPortEnv specifies the name of the env variable that contains driver port.
 	DriverPortEnv = "DRIVER_PORT"
 
+	// KillAfterEnv specifies the name of the env variable that sets the allowed
+	// response time for a pod after timeout.
+	KillAfterEnv = "KILL_AFTER"
+
+	// NonProxiedTargetStringEnv specifies the target string that PSM test client send
+	// traffic to.
+	NonProxiedTargetStringEnv = "NON_PROXIED_TARGET_STRING"
```
[the remark this answers is not in the document]
Comment (wanlin31) [reply config/constants.go:75]: Great suggestion, I will update them wherever applies.

File: containers/runtime/xds/config/config.go (hunk 6 above)
Comment (wanlin31) [reply containers/runtime/xds/config/config.go:461]: I make it `IncludeSocketListenerOnly`. Thanks.

File: containers/runtime/xds/logger.go (hunk 7 above)
Comment (wanlin31) [reply containers/runtime/xds/logger.go:18]: I updated

File: containers/init/ready/ready.go (hunk 2 above)
Comment (wanlin31) [reply containers/init/ready/ready.go:284]: My understanding is since it was separated to a different go-routine, it would not stop the ready container from moving forward. The endpoint (I should have updated the name, I will fix it) update server in the xds server is also in a separate go-routine, I don't think the failure would do anything. Not able to quit it may just cause it to be continuously running.

File: kubehelpers/configMap.go (hunk 8 above)
Comment (wanlin31) [reply kubehelpers/configMap.go:26]: Right. I actually want to discuss with you in this. Do you expect we run multiple scenarios in the same LoadTest? Currently I only support update the server port when there is only one scenario.

File: proto/testupdater/test_updater.proto (hunk 10 above)
Comment (wanlin31) [reply proto/testupdater/test_updater.proto:25]: I will update it. :)

File: containers/runtime/xds/logger.go (hunk 7 above)
Comment (wanlin31) [reply containers/runtime/xds/logger.go:18]: It does.

File: status/status.go (hunk 11 above)
Comment (wanlin31) [reply status/status.go:57]: As discussed, I will have a separate PR to address this.

File: containers/init/ready/ready.go (hunk 1 above)
Comment (wanlin31) [reply containers/init/ready/ready.go:260]: I forgot to add an input argument here, we can probably ask it to check if there is specific ENV, such as "QUITQUITQUIT", is so we quit the update server. Do you see a need that the user might not want to quit the test update server? If not we can probably remove that parameter and quit the test update server always.

File: api/v1/loadtest_types.go [hunk 13]
```diff
@@ -265,6 +319,14 @@ type Client struct {
 	// +optional
 	Build *Build `json:"build,omitempty"`
 
+	// Sidecar describes the Sidecar container, which is the runtime of
+	// the proxy for the test associated with a proxy.
+	Sidecar *Sidecar `json:"sidecar,omitempty"`
```
[the remark this answers is not in the document]
Comment (wanlin31) [reply api/v1/loadtest_types.go:324]: This is a great point, I always thought we need to wrap the container with something, but seems like all the required fields are in the [corev1.Container](https://github.com/kubernetes/api/blob/v0.20.2/core/v1/types.go#L2189). Hi @paulosjca, do you remember why we need to wrap the clone, build and run containers instead of use the container directly?

[the remark this answers is not in the document]
Comment (wanlin31) [reply api/v1/loadtest_types.go:324]: As I discussed offline with @paulosjca, we may change the loadTest configuration definition to be more generic (make all run) before this #255, and the xdsServer and sidecar maybe just Containers.

File: containers/init/ready/ready.go (hunk 1 above)
Comment (gnossen) [reply containers/init/ready/ready.go:260]: > Do you see a need that the user might not want to quit the test update server?
Actually, I don't understand why the update server needs to have a remote termination feature at all. Can you please help me understand this?

File: containers/init/ready/ready.go (hunk 2 above)
Comment (gnossen) [reply containers/init/ready/ready.go:284]: Closing this thread in favor of [this one](https://github.com/grpc/test-infra/pull/255#discussion_r798138526) since they're essentially the same question.

File: proto/testupdater/test_updater.proto (hunk 10 above)
Comment (gnossen) [reply proto/testupdater/test_updater.proto:25]: It's not important. Don't worry about the name change.

File: kubehelpers/configMap.go (hunk 8 above)
Comment (gnossen) [reply kubehelpers/configMap.go:26]: I think a single scenario at a time will be fine for now.

**Review by gnossen — APPROVED** [review]
